### DISCUSSION TITLE
Auto Restart (Windows) CMD Batch?

### DISCUSSION THREAD

**Discussion by Legacy-TacticalGamingInteractive** [SnailyCAD/snaily-cadv4#1537] [Q&A] [answered]
### Describe the feature

Wondering if you know how to code into the CMD window process to auto restart if a crash is detected. I cannot keep restarting manually once a day its getting too hectic.

```
[1]   errors: {
[1]     'violations.cky62eiek4674aotmih7f36yc.fine': { message: 'fine_invalidDataReceived', data: [Object] }
[1]   }
[1] }
[1] uncaughtException captured: BAD_REQUEST(400): badRequest
[1] yarn workspace @snailycad/api generate && yarn workspace @snailycad/api start exited with code 1
[0] yarn workspace @snailycad/client start exited with code 1
^CTerminate batch job (Y/N)? Y

C:\nginx\html\snaily-cadv4>yarn run concurrently "yarn workspace @snailycad/client start" "yarn workspace @snailycad/api generate && yarn workspace @snailycad/api start"
```

For example if the script has detected:
```
[1] yarn workspace @snailycad/api generate && yarn workspace @snailycad/api start exited with code 1
[0] yarn workspace @snailycad/client start exited with code 1
```

then it will automatically attempt to run:
```
yarn run concurrently "yarn workspace @snailycad/client start" "yarn workspace @snailycad/api generate && yarn workspace @snailycad/api start"
```

I'm not sure how to implement this myself.

### Additional Context

Daily crashing is becoming annoying to deal with and my CAD is down for hours sometimes before someone tells me.

**Comment by casperiv0** [comment, marked as answer]
Check out `pm2`. This will run the CAD as a service in the background, should be supported on all platforms: https://docs.snailycad.org/docs/guides/how-to-start-with-pm2

**Comment by Legacy-TacticalGamingInteractive** [comment]
![image](https://user-images.githubusercontent.com/25211271/219948970-9f7457c4-613d-41e1-af4e-58703e1f730f.png)

it says its stopped
but yet its not stopped... the cad is running
did i do something wrong?

**Comment by casperiv0** [comment]
yeah, make sure to **stop** the current CAD process, then start it via PM2. Otherwise you'll have 2 CAD processes

**Reply by Legacy-TacticalGamingInteractive** [reply]
So I should go to the window with PM2 running and then start it CAD?

**Reply by casperiv0** [reply]
no, you should only start via PM2.

**Reply by Legacy-TacticalGamingInteractive** [reply]
oh just like this? thats it?

![image](https://user-images.githubusercontent.com/25211271/219950920-64136dfa-ce0c-4116-8317-4d828567256e.png)

**Reply by casperiv0** [reply]
Yes. Then you can use `pm2 logs` to view the logs. Make sure to also run `pm2 startup && pm2 save` to save the configuration and start SnailyCAD on startup.

**Reply by Legacy-TacticalGamingInteractive** [reply]
![image](https://user-images.githubusercontent.com/25211271/219951116-de3d0fb0-009d-4ae1-9106-4b497b631bd8.png)

**Reply by casperiv0** [reply]
try `pm2 delete all && pm2 start npm --name SnailyCADv4 -- run start`

**Reply by Legacy-TacticalGamingInteractive** [reply]
it just keeps giving me errors or even if i get it to say online it still gives me a 502 cad error so for now im giving up on it ill try later again i dont know

when i run that command above it lists PID 0 but its stopped

**Reply by casperiv0** [reply]
not sure about that then. Try different ways, there must be more ways to use this or other tools
